src/Compilers/Core/Portable/IReferenceOrISignature.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Runtime.CompilerServices;
+using Microsoft.Cci;
+using Microsoft.CodeAnalysis.Symbols;
+
+namespace Microsoft.CodeAnalysis
+{
+    // These types are to enable fast-path devirtualization in the Jit. Dictionary<K, V>, HashTable<T>
+    // and ConcurrentDictionary<K, V> will devirtualize (and potentially inline) the IEquatable<T>.Equals
+    // method for a struct when the Comparer is unspecified in .NET Core, .NET 5; whereas specifying
+    // a Comparer will make .Equals and GetHashcode slower interface calls.
+
+    /// <summary>
+    /// Used to devirtualize Dictionary/HashSet for EqualityComparer{T}.Default
+    /// </summary>
+    internal readonly struct IReferenceOrISignatureEquivalent : IEquatable<IReferenceOrISignatureEquivalent>
+    {
+        private readonly object? _item;
```

Comment:
Why is this field nullable? And why is `Equals` optimized so that checking for `null` is fast? Specifically, I think the changes in https://github.com/svick/roslyn/commit/5ebd2af2ef98d33b66c5be3089aa758fcab286c8 should be safe.

Reply:
Not sure what the nullablity approach is around with struct types and `default`?

Though it should never be created as `default`

Reply:
Changed to non-nullable